src/code-block/files.ts

```diff
@@ -0,0 +1,78 @@
+import {
+  App,
+  TFolder,
+  TFile,
+  FileStats,
+  TAbstractFile,
+  Vault,
+  MarkdownPostProcessorContext,
+} from "obsidian";
+import { Settings } from "~/code-block/settings";
+import renderError from "~/code-block/errors";
+
+type WVault = Vault & {
+  getAbstractFileByPathInsensitive?: null | ((path: string) => TAbstractFile | null);
+};
+
+const VALID_EXTENSIONS = ["jpeg", "jpg", "gif", "png", "webp", "tiff", "tif", "md"];
+
+const getChildren = (children: TAbstractFile[], recursive = false) =>
+  children.reduce((acc, v) => {
+    if (recursive && v instanceof TFolder) {
+      acc = acc.concat(getChildren(v.children, true));
+    } else if (v instanceof TFile) {
+      acc.push(v);
+    }
+    return acc;
+  }, [] as Array<TFile>);
+
+const getPath = (app: App, path: string) => {
+  const { vault } = app;
+  const wVault: WVault = vault;
+  if (wVault.getAbstractFileByPathInsensitive) {
+    return wVault.getAbstractFileByPathInsensitive(path);
+  }
+  return vault.getAbstractFileByPath(path);
+};
+
+const getFileList = (
+  app: App,
+  ctx: MarkdownPostProcessorContext,
+  container: HTMLElement,
+  settings: Settings,
+) => {
+  // retrieve a list of files in the settings.path
+  // and if specified by settings.recursive fetch the files
+  // in all the subfolders.
+  let files: TFile[];
+  const folder = getPath(app, settings.path);
+  if (folder instanceof TFolder) {
+    files = getChildren(folder.children, settings.recursive);
+  } else {
+    const error = "The folder doesn't exist, or it's empty!";
+    renderError(container, error);
+    throw new Error(error);
+  }
```

Comment:
[🔧](https://github.com/erikthedeveloper/code-review-emoji-guide#:~:text=:wrench:) You can avoid mutating the `files` variable by using an early-return for the error path

(btw small typo on `it's empty` → `is empty`)

```suggestion
  const folder = getPath(app, settings.path);
  if (!(folder instanceof TFolder)) {
    const error = "The folder doesn't exist, or is empty!";
    renderError(container, error);
    throw new Error(error);
  } 
  const files = getChildren(folder.children, settings.recursive);
```